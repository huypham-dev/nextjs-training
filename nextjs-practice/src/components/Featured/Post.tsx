import { memo } from "react";

import { PostType } from "types";
import Link from "next/link";
import Image from "next/image";
import isEqual from "react-fast-compare";

type Props = {
  post: PostType;
};

const Post = ({ post }: Props) => {
  const { id, title, image, view } = post;

  return (
    <div className="w-full md:w-1/2 lg:w-1/3 px-6 p-5">
      <div className="bg- p-4 rounded-md shadow-xl border-gray-200 border-2">
        <h3 className="text-md font-extrabold tracking-widest mb-2">{title}</h3>
        <div className="w-full text-center">
          <Image
            className="rounded-lg"
            src={image}
            alt="article image"
            width={250}
            height={150}
            layout="responsive"
          />
        </div>
        <p className="text-xs text-gray-500 mb-4">
          {view} <i>views</i>
        </p>
        <Link href={`/post/${id}`}>
          <a className="inline-block border-2 border-black py-3 px-8 uppercase font-display font-thin tracking-wide text-xs hover:bg-yellow-400 hover:border-yellow-400">
            Read more
          </a>
        </Link>
      </div>
    </div>
  );
};

export default memo(Post, isEqual);
