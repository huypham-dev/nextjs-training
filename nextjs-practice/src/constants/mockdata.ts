export const CATEGORIES = [
  {
    id: "1",
    name: "mobile",
    image:
      "https://images.prismic.io/nextjsexample/7853f863-6a12-4312-a6da-f930437f6d62_mobile.png",
  },
  {
    id: "2",
    name: "internet",
    image:
      "https://images.prismic.io/nextjsexample/cb71ce70-bb03-442b-99e8-a318255b81f3_3579187.png",
  },
];

export const ARTICLES = [
  {
    id: "1",
    createdAt: "2021-03-01T12:59:16.855Z",
    title: "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
    author: "author 1",
    image:
      "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg",
    category: "internet",
    content:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.1",
    slug: "slug 1",
    view: 90,
  },
  {
    id: "2",
    createdAt: "2021-03-01T12:59:16.855Z",
    title: "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
    author: "author 1",
    image:
      "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg",
    category: "internet",
    content:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.1",
    slug: "slug 1",
    view: 90,
  },
];
