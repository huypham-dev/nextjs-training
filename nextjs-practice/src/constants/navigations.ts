export const navigations = [
  { label: "Home", path: "/" },
  { label: "Blog", path: "/blog" },
  { label: "Contact", path: "/contact" },
];
