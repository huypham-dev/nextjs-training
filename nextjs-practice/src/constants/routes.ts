export const ROUTES = {
  home: {
    path: "/",
  },
  blog: {
    path: "/blog",
    withId: (id: string) => `/blog/${id}`,
  },
  post: {
    path: "/post",
    withId: (id: string) => `/post/${id}`,
  },
  contact: {
    path: "/",
  },
};
