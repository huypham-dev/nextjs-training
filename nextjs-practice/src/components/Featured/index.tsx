import { memo } from "react";
import Post from "./Post";
import { PostType } from "types";
import isEqual from "react-fast-compare";

type Props = {
  featuredPosts: PostType[];
};

const Featured = ({ featuredPosts }: Props) => (
  <div className="mx-auto max-w-6xl bg-white shadow-xl mb-10">
    <div className="max-w-5xl mx-auto pt-10 p-8 flex">
      <h3 className="font-extrabold text-5xl md:text-6xl italic leading-none max-w-xl">
        Featured Posts
      </h3>
    </div>
    <div className="container mx-auto flex flex-wrap pb-20 px-10">
      {featuredPosts.map((item) => (
        <Post key={item.id} post={item} />
      ))}
    </div>
  </div>
);

export default memo(Featured, isEqual);
