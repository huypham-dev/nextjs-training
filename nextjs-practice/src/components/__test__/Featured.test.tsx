import { mockArticles } from "constants/mockdata";
import { mount, shallow } from "enzyme";
import Featured from "../Featured";
import Post from "../Featured/Post";

describe("Featured", () => {
  it("Snapshot", () => {
    const component = shallow(<Featured featuredPosts={mockArticles} />);

    expect(component).toMatchSnapshot();
  });

  it("Render component correctly with props", () => {
    const component = mount(<Featured featuredPosts={mockArticles} />);
    const posts = component.find(Post).length;
    const title = component.find("h3").at(0).text();

    expect(posts).toBe(2);
    expect(title).toBe("Featured Posts");
  });
});
