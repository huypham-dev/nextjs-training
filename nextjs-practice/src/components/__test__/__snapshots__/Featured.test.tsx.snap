// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Featured Snapshot 1`] = `
<div
  className="mx-auto max-w-6xl bg-white shadow-xl mb-10"
>
  <div
    className="max-w-5xl mx-auto pt-10 p-8 flex"
  >
    <h3
      className="font-extrabold text-5xl md:text-6xl italic leading-none max-w-xl"
    >
      Featured Posts
    </h3>
  </div>
  <div
    className="container mx-auto flex flex-wrap pb-20 px-10"
  >
    <Memo(Post)
      key="1"
      post={
        Object {
          "author": "author 1",
          "category": "internet",
          "content": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.1",
          "createdAt": "2021-03-01T12:59:16.855Z",
          "id": "1",
          "image": "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg",
          "slug": "slug 1",
          "title": "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
          "view": 90,
        }
      }
    />
    <Memo(Post)
      key="2"
      post={
        Object {
          "author": "author 1",
          "category": "internet",
          "content": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.1",
          "createdAt": "2021-03-01T12:59:16.855Z",
          "id": "2",
          "image": "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg",
          "slug": "slug 1",
          "title": "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
          "view": 90,
        }
      }
    />
  </div>
</div>
`;
